pip/private/constraints/abi/macro.bzl

```diff
@@ -0,0 +1,122 @@
+load("@bazel_skylib//lib:selects.bzl", "selects")
+load("//pip/private/constraints:defs.bzl", "MAJORS", "MINORS", "INTERPRETERS", "FLAGS")
+
+# FIXME: Where does abi 2/3/4 fit in here?
+# FIXME: Where do ABI feature flags fit in here?
+def generate():
+    """
+    Lay down `py3`, `py312`, `cp3`, `cp312` etc and critically `any`.
+
+    The interpretation is a bit tricky because `cp`
+    """
+
+    # FIXME: Is there a better/worse way to do this?
+    selects.config_setting_group(
+        name = "none",
+        match_all = [
+            "//conditions:default",
+        ]
+    )
+
+    native.constraint_setting(
+        name = "feature_pydebug",
+        default_constraint_value = ":pydebug_disabled",
+    )
+    native.constraint_value(
+        name = "pydebug_enabled",
+        constraint_setting = ":feature_pydebug",
+    )
+    native.constraint_value(
+        name = "pydebug_disabled",
+        constraint_setting = ":feature_pydebug",
+    )
+
+    native.constraint_setting(
+        name = "feature_pymalloc",
+        default_constraint_value = ":pymalloc_disabled",
+    )
+    native.constraint_value(
+        name = "pymalloc_enabled",
+        constraint_setting = ":feature_pymalloc",
+    )
+    native.constraint_value(
+        name = "pymalloc_disabled",
+        constraint_setting = ":feature_pymalloc",
+    )
+
+    native.constraint_setting(
+        name = "feature_freethreading",
+        default_constraint_value = ":freethreading_disabled",
+    )
+    native.constraint_value(
+        name = "freethreading_enabled",
+        constraint_setting = ":feature_freethreading",
+    )
+    native.constraint_value(
+        name = "freethreading_disabled",
+        constraint_setting = ":feature_freethreading",
+    )
+
+    native.constraint_setting(
+        name = "feature_wide_unicode",
+        default_constraint_value = ":wide_unicode_disabled",
+    )
+    native.constraint_value(
+        name = "wide_unicode_enabled",
+        constraint_setting = ":feature_wide_unicode",
+    )
+    native.constraint_value(
+        name = "wide_unicode_disabled",
+        constraint_setting = ":feature_wide_unicode",
+    )
+
+    native.alias(
+        name = "abi3",
+        actual = "is_py33"
+    )
+
+    for interpreter in INTERPRETERS:
+        for major in MAJORS:
+            # selects.config_setting_group(
+            #     name = "{}{}".format(interpreter, major),
+            #     match_all = [
+            #         "//pip/private/constraints/python/interpreter:{}".format(interpreter),
+            #         "//pip/private/constraints/python/major:{}".format(major),
+            #     ]
+            # )
+
+            for minor in MINORS:
+                selects.config_setting_group(
+                    name = "is_{}{}{}".format(interpreter, major, minor),
```

Comment:
generally `%s` is faster than `.format` because it avoids a function call and has simpler (faster) arg conversion. Also if you have only a single substitution it helps even more because you don't even need to construct a tuple. Up to you where to use it, but at least for the functions in the extension that are called repeatedly I would strongly consider it. (Honestly I pretty much always use it unless I have a complex/multiline template with 4+ args and/or using an arg multiple times)